js/views/modals/purchase/Payment.js

```diff
@@ -0,0 +1,175 @@
+import $ from 'jquery';
+import app from '../../../app';
+import loadTemplate from '../../../utils/loadTemplate';
+import { convertAndFormatCurrency, integerToDecimal } from '../../../utils/currency';
+import { getSocket } from '../../../utils/serverConnect';
+import BaseVw from '../../baseVw';
+import ConfirmWallet from './ConfirmWallet';
+import qr from 'qr-encode';
+import { clipboard, remote } from 'electron';
+import Purchase from '../../../models/purchase/Purchase';
+import { spend } from '../../../models/wallet/Spend';
+import { openSimpleMessage } from '../../modals/SimpleMessage';
+
+export default class extends BaseVw {
+  constructor(options = {}) {
+    if (!options.model || !options.model instanceof Purchase) {
+      throw new Error('Please provide a purchase model.');
+    }
+
+    super(options);
+    this.options = options;
+
+    this.boundOnDocClick = this.onDocumentClick.bind(this);
+    $(document).on('click', this.boundOnDocClick);
+
+    const serverSocket = getSocket();
+    if (serverSocket) {
+      this.listenTo(serverSocket, 'message', e => {
+        // listen for a payment socket message, to react to payments from all sources
+        if (e.jsonData.notification && e.jsonData.notification.payment) {
+          const payment = e.jsonData.notification.payment;
+          if (integerToDecimal(payment.fundingTotal, true) >= this.model.get('amount') &&
+            payment.orderId === this.model.get('orderId')) {
+            this.trigger('walletPaymentComplete', payment);
+          }
+        }
+      });
+    }
+  }
+
+  className() {
+    return 'pending';
+  }
+
+  events() {
+    return {
+      'click .js-payFromWallet': 'clickPayFromWallet',
+      'click .js-payFromAlt': 'clickPayFromAlt',
+      'click .js-copyAmount': 'copyAmount',
+      'click .js-copyAddress': 'copyAddress',
+    };
+  }
+
+  onDocumentClick(e) {
+    if (!($.contains(this.$confirmWallet[0], e.target))) {
+      this.$confirmWallet.addClass('hide');
+    }
+  }
+
+  clickPayFromWallet(e) {
+    e.stopPropagation();
+    this.$confirmWallet.removeClass('hide');
+  }
+
+  walletCancel() {
+    this.$confirmWallet.addClass('hide');
+  }
+
+  walletConfirm() {
+    this.$confirmWalletConfirm.addClass('processing');
+
+    spend({
+      address: this.model.get('paymentAddress'),
+      amount: this.model.get('amount'),
+      currency: 'BTC',
+    })
+      .fail(jqXhr => {
+        openSimpleMessage(app.polyglot.t('purchase.errors.paymentFailed'),
+          jqXhr.responseJSON && jqXhr.responseJSON.reason || '');
+      })
+      .always(() => {
```

Comment:
I think it's ok to not abort this call on remove(), since we want the wallet to reflect the payment. Also, it's probably not bad for the user to see any errors if the Purchase fails on the server.

But... I do think we probably want to throw a `if removed() return` in the always handler. What's in there is pretty benign if executed after the modal closes, but for future proofing it wouldn't hurt.